### PR TITLE
Add extra poll() call to work around libcurl issue

### LINKED ISSUE #9751: HttpContent Stream Read() and ReadAsync() consume 100% CPU on Scientific Linux 7.2
Running 1.0.0-preview2-003121 on Scientific Linux release 7.2 (Nitrogen), which `dotnet --info` identifies as (OS Name:     rhel, OS Version:  7.2, OS Platform: Linux, RID:         rhel.7.2-x64).

In order to read from a SSE EventSource (which is basically just a HTTP GET that doesn't close immediately), my code currently:

1. Creates a `HttpClient`
2. Sets up a `HttpRequestMessage` and uses it in a `SendAsync()` on the `HttpClient` to get a `HttpResponseMessage`
3. awaits `HttpResponseMessage.Content.ReadAsStreamAsync()`
4. Enters into a loop calling `ReadAsync()` on the `Stream`, which is then fed into a decoder.

The issue is that on RHEL 7.2, `await ReadAsync()` blocks (as expected), but with 100% CPU used by the `dotnet` process. This also occurs if the `ReadAsync()` is replaced by the standard blocking `Read()`. To my untrained eye, it appears to be spin waiting or spin locking, or something in that vain. 

I have tested and confirmed that the issue does not manifest itself on Windows 10 (build 14372) or Ubuntu 14.04.

I have attached a small example project that reproduces the issue. It is unfortunately slightly cumbersome to run, since an event source must be hosted for it to read, and I can't find an existing example event source server to point it to.

[RHELBugTest.zip](https://github.com/dotnet/corefx/files/338889/RHELBugTest.zip)

### PR DESCRIPTION
**Pull request by ericeil**
libcurl versions prior to 7.32.0 have a bug in `curl_multi_wait`: the `revents` fields for the "extra" file descriptors are not populated with the relevant events that are signaled.  We rely on this info to know when to read data from a pipe, to reset those events, and so under some circumstances we can end up with a pipe that is always readable, leading to `curl_multi_wait` always returning immediately.  This results in 100% CPU usage for some scenarios.

This bug is fixed in libcurl 7.32.0 with [this commit](https://github.com/curl/curl/commit/6d30f8ebed34e7276c2a59ee20d466bff17fee56).  However, this version is not available on some supported Linux distros, so a workaround is also needed.  We simply add a manual `poll` call in the case where we're running on a libcurl version without the fix.

Fixes #9751.

cc: @stephentoub, @crozone, @CRCinAU, @ppanyukov, @TingluoHuang

### CODE REVIEW

**Comment by ericeil** [comment]
@dotnet-bot test Innerloop Windows_NT Debug Build and Test please
(timed out in a networking test on Windows, unrelated to this change)

**Comment by CRCinAU** [comment]
I'm curious. How does this interact with vendors like RH that backport these patches?

As a test, I have the preview of the libcurl packages that will be released with the RHEL 7.3 refresh at:
http://au1.mirror.crc.id.au/repo/el7-testing/x86_64/

File: src/Native/Unix/System.Net.Http.Native/pal_multi.cpp [hunk 1]
```diff
@@ -69,8 +70,31 @@ extern "C" int32_t HttpNative_MultiWait(CURLM* multiHandle,
...
+    else
+    {
+        *isTimeout = false;
+
+        //
+        // Prior to libcurl version 7.32.0, the revents field was not returned properly for "extra" file descriptors
+        // passed to curl_multi_wait.  See https://github.com/dotnet/corefx/issues/9751.  So if we have a libcurl
+        // prior to that version, we need to do our own poll to get the status of the extra file descriptor.
+        //
+        if (curl_version_info(CURLVERSION_NOW)->version_num >= 0x073200)
```
Comment (stephentoub) [inline src/Native/Unix/System.Net.Http.Native/pal_multi.cpp]: Any sense for what the cost of this call is?  The version of curl being used shouldn't change, so it seems like something we could look up once and cache if curl_version_info has any measurable expense to it.

File: src/Native/Unix/System.Net.Http.Native/pal_multi.cpp [hunk 2]
```diff
@@ -69,8 +70,31 @@ extern "C" int32_t HttpNative_MultiWait(CURLM* multiHandle,
...
+        {
+            *isExtraFileDescriptorActive = (extraFds.revents & CURL_WAIT_POLLIN) != 0;
+        }
+        else
+        {
+            pollfd pfd = { .fd = ToFileDescriptor(extraFileDescriptor),.events = POLLIN,.revents = 0 };
+            poll(&pfd, 1, 0);
```
Comment (stephentoub) [inline src/Native/Unix/System.Net.Http.Native/pal_multi.cpp]: I assume we're ignoring any poll result/error on purpose? Assuming so, a comment to that effect and why it's ok would be good.

**Comment by stephentoub** [comment]
> How does this interact with vendors like RH that backport these patches?

It seems like, worst case, we'd do any extra unnecessary-but-not-harmful poll.  @ericeil, is that correct?

File: src/Native/Unix/System.Net.Http.Native/pal_multi.cpp (hunk 1 above)
Comment (ericeil) [inline src/Native/Unix/System.Net.Http.Native/pal_multi.cpp]: I originally wrote the code to cache the `curl_version_info` result, but it turns out that `curl_version_info` already does that internally.  Calling this should be very cheap.

**Comment by ericeil** [comment]
> > How does this interact with vendors like RH that backport these patches?
>
> It seems like, worst case, we'd do any extra unnecessary-but-not-harmful poll.  @ericeil, is that correct?

Yes, that's correct.  The scenarios are:

1. Unpatched old libcurl versions will cause us to invoke an extra poll, adding a little overhead, but eliminating the 100% CPU usage problem.
2. Patched old libcurl versions will still have the extra poll, so will have the same extra overhead as the unpatched libcurl.
3. Once libcurl is upgraded to a newer version, the overhead goes away.

File: src/Native/Unix/System.Net.Http.Native/pal_multi.cpp (hunk 2 above)
Comment (ericeil) [inline src/Native/Unix/System.Net.Http.Native/pal_multi.cpp]: Added a comment about this.

**Comment by stephentoub** [comment]
Test Innerloop CentOS7.1 Debug Build and Test please

**Pull request merged**